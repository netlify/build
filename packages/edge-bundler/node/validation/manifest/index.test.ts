import chalk from 'chalk'
import { test, expect, describe } from 'vitest'

import { validateManifest, ManifestValidationError } from './index.js'

// We need to disable all color outputs for the tests as they are different on different platforms, CI, etc.
// This only works if this is the same instance of chalk that better-ajv-errors uses
chalk.level = 0

// Factory so we have a new object per test
const getBaseManifest = (): Record<string, any> => ({
  bundles: [
    {
      asset: 'f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js',
      format: 'js',
    },
  ],
  routes: [
    {
      name: 'name',
      function: 'hello',
      pattern: '^/hello/?$',
      generator: '@netlify/fake-plugin@1.0.0',
    },
  ],
  post_cache_routes: [
    {
      name: 'name',
      function: 'hello',
      pattern: '^/hello/?$',
      generator: '@netlify/fake-plugin@1.0.0',
    },
  ],
  layers: [
    {
      flag: 'flag',
      name: 'name',
      local: 'local',
    },
  ],
  bundler_version: '1.6.0',
})

test('should not throw on valid manifest', () => {
  expect(() => validateManifest(getBaseManifest())).not.toThrowError()
})

test('should throw ManifestValidationError with correct message', () => {
  expect(() => validateManifest('manifest')).toThrowError(ManifestValidationError)
  expect(() => validateManifest('manifest')).toThrowError(/^Validation of Edge Functions manifest failed/)
})

test('should throw ManifestValidationError with customErrorInfo', () => {
  try {
    validateManifest('manifest')
  } catch (error) {
    expect(error).toBeInstanceOf(ManifestValidationError)

    const { customErrorInfo } = error as ManifestValidationError
    expect(customErrorInfo).toBeDefined()
    expect(customErrorInfo.type).toBe('functionsBundling')
    return
  }

  expect.fail('should have thrown')
})

test('should throw on additional property on root level', () => {
  const manifest = getBaseManifest()
  manifest.foo = 'bar'

  expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
})

test('should show multiple errors', () => {
  const manifest = getBaseManifest()
  manifest.foo = 'bar'
  manifest.baz = 'bar'

  expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
})

describe('bundle', () => {
  test('should throw on additional property in bundle', () => {
    const manifest = getBaseManifest()
    manifest.bundles[0].foo = 'bar'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing asset', () => {
    const manifest = getBaseManifest()
    delete manifest.bundles[0].asset

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing format', () => {
    const manifest = getBaseManifest()
    delete manifest.bundles[0].format

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on invalid format', () => {
    const manifest = getBaseManifest()
    manifest.bundles[0].format = 'foo'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})

describe('route', () => {
  test('should throw on additional property', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].foo = 'bar'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on invalid pattern', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].pattern = '/^/hello/?$/'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing function', () => {
    const manifest = getBaseManifest()
    delete manifest.routes[0].function

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing pattern', () => {
    const manifest = getBaseManifest()
    delete manifest.routes[0].pattern

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})

// No tests for post_cache_routes as schema shared with routes

describe('layers', () => {
  test('should throw on additional property', () => {
    const manifest = getBaseManifest()
    manifest.layers[0].foo = 'bar'

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing name', () => {
    const manifest = getBaseManifest()
    delete manifest.layers[0].name

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on missing flag', () => {
    const manifest = getBaseManifest()
    delete manifest.layers[0].flag

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})

describe('import map URL', () => {
  test('should accept string value', () => {
    const manifest = getBaseManifest()
    manifest.import_map = 'file:///root/.netlify/edge-functions-dist/import_map.json'

    expect(() => validateManifest(manifest)).not.toThrowError()
  })

  test('should throw on wrong type', () => {
    const manifest = getBaseManifest()
    manifest.import_map = ['file:///root/.netlify/edge-functions-dist/import_map.json']

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })
})

describe('route headers', () => {
  test('should accept valid headers with exists matcher', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'exists',
      },
    }

    expect(() => validateManifest(manifest)).not.toThrowError()
  })

  test('should accept valid headers with missing matcher', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'missing',
      },
    }

    expect(() => validateManifest(manifest)).not.toThrowError()
  })

  test('should accept valid headers with regex matcher and pattern', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'regex',
        pattern: 'Bearer .+',
      },
    }

    expect(() => validateManifest(manifest)).not.toThrowError()
  })

  test('should throw on missing matcher property', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        pattern: '^Bearer .+$',
      },
    }

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on invalid matcher value', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'invalid',
      },
    }

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw when matcher is regex but pattern is missing', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'regex',
      },
    }

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on invalid pattern format', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'regex',
        pattern: /^Bearer .+/,
      },
    }

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should throw on additional property in headers', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-custom-header': {
        matcher: 'exists',
        foo: 'bar',
      },
    }

    expect(() => validateManifest(manifest)).toThrowErrorMatchingSnapshot()
  })

  test('should accept multiple headers with different matchers', () => {
    const manifest = getBaseManifest()
    manifest.routes[0].headers = {
      'x-exists-header': {
        matcher: 'exists',
      },
      'x-missing-header': {
        matcher: 'missing',
      },
      authorization: {
        matcher: 'regex',
        pattern: '^Bearer [a-zA-Z0-9]+$',
      },
    }

    expect(() => validateManifest(manifest)).not.toThrowError()
  })
})
