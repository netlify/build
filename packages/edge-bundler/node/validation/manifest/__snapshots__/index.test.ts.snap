// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bundle > should throw on additional property in bundle 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  4 |       \\"asset\\": \\"f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js\\",
  5 |       \\"format\\": \\"js\\",
> 6 |       \\"foo\\": \\"bar\\"
    |       ^^^^^ 😲  foo is not expected to be here!
  7 |     }
  8 |   ],
  9 |   \\"routes\\": ["
`;

exports[`bundle > should throw on invalid format 1`] = `
"Validation of Edge Functions manifest failed
ENUM must be equal to one of the allowed values
(eszip2, js)

  3 |     {
  4 |       \\"asset\\": \\"f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js\\",
> 5 |       \\"format\\": \\"foo\\"
    |                 ^^^^^ 👈🏽  Unexpected value, should be equal to one of the allowed values
  6 |     }
  7 |   ],
  8 |   \\"routes\\": ["
`;

exports[`bundle > should throw on missing asset 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'asset'

  1 | {
  2 |   \\"bundles\\": [
> 3 |     {
    |     ^ ☹️  asset is missing here!
  4 |       \\"format\\": \\"js\\"
  5 |     }
  6 |   ],"
`;

exports[`bundle > should throw on missing format 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'format'

  1 | {
  2 |   \\"bundles\\": [
> 3 |     {
    |     ^ ☹️  format is missing here!
  4 |       \\"asset\\": \\"f35baff44129a8f6be7db68590b2efd86ed4ba29000e2edbcaddc5d620d7d043.js\\"
  5 |     }
  6 |   ],"
`;

exports[`import map URL > should throw on wrong type 1`] = `
"Validation of Edge Functions manifest failed
TYPE must be string

  30 |   ],
  31 |   \\"bundler_version\\": \\"1.6.0\\",
> 32 |   \\"import_map\\": [
     |                 ^
> 33 |     \\"file:///root/.netlify/edge-functions-dist/import_map.json\\"
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 34 |   ]
     | ^^^^ 👈🏽  type must be string
  35 | }"
`;

exports[`layers > should throw on additional property 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  27 |       \\"name\\": \\"name\\",
  28 |       \\"local\\": \\"local\\",
> 29 |       \\"foo\\": \\"bar\\"
     |       ^^^^^ 😲  foo is not expected to be here!
  30 |     }
  31 |   ],
  32 |   \\"bundler_version\\": \\"1.6.0\\""
`;

exports[`layers > should throw on missing flag 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'flag'

  23 |   ],
  24 |   \\"layers\\": [
> 25 |     {
     |     ^ ☹️  flag is missing here!
  26 |       \\"name\\": \\"name\\",
  27 |       \\"local\\": \\"local\\"
  28 |     }"
`;

exports[`layers > should throw on missing name 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'name'

  23 |   ],
  24 |   \\"layers\\": [
> 25 |     {
     |     ^ ☹️  name is missing here!
  26 |       \\"flag\\": \\"flag\\",
  27 |       \\"local\\": \\"local\\"
  28 |     }"
`;

exports[`route > should throw on additional property 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  12 |       \\"pattern\\": \\"^/hello/?$\\",
  13 |       \\"generator\\": \\"@netlify/fake-plugin@1.0.0\\",
> 14 |       \\"foo\\": \\"bar\\"
     |       ^^^^^ 😲  foo is not expected to be here!
  15 |     }
  16 |   ],
  17 |   \\"post_cache_routes\\": ["
`;

exports[`route > should throw on invalid pattern 1`] = `
"Validation of Edge Functions manifest failed
FORMAT pattern must be a regex that starts with ^ and ends with $ (e.g. ^/blog/[d]{4}$)

  10 |       \\"name\\": \\"name\\",
  11 |       \\"function\\": \\"hello\\",
> 12 |       \\"pattern\\": \\"/^/hello/?$/\\",
     |                  ^^^^^^^^^^^^^^ 👈🏽  format pattern must be a regex that starts with ^ and ends with $ (e.g. ^/blog/[d]{4}$)
  13 |       \\"generator\\": \\"@netlify/fake-plugin@1.0.0\\"
  14 |     }
  15 |   ],"
`;

exports[`route > should throw on missing function 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'function'

   7 |   ],
   8 |   \\"routes\\": [
>  9 |     {
     |     ^ ☹️  function is missing here!
  10 |       \\"name\\": \\"name\\",
  11 |       \\"pattern\\": \\"^/hello/?$\\",
  12 |       \\"generator\\": \\"@netlify/fake-plugin@1.0.0\\""
`;

exports[`route > should throw on missing pattern 1`] = `
"Validation of Edge Functions manifest failed
REQUIRED must have required property 'pattern'

   7 |   ],
   8 |   \\"routes\\": [
>  9 |     {
     |     ^ ☹️  pattern is missing here!
  10 |       \\"name\\": \\"name\\",
  11 |       \\"function\\": \\"hello\\",
  12 |       \\"generator\\": \\"@netlify/fake-plugin@1.0.0\\""
`;

exports[`should show multiple errors 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  30 |   ],
  31 |   \\"bundler_version\\": \\"1.6.0\\",
> 32 |   \\"foo\\": \\"bar\\",
     |   ^^^^^ 😲  foo is not expected to be here!
  33 |   \\"baz\\": \\"bar\\"
  34 | }

ADDTIONAL PROPERTY must NOT have additional properties

  31 |   \\"bundler_version\\": \\"1.6.0\\",
  32 |   \\"foo\\": \\"bar\\",
> 33 |   \\"baz\\": \\"bar\\"
     |   ^^^^^ 😲  baz is not expected to be here!
  34 | }"
`;

exports[`should throw on additional property on root level 1`] = `
"Validation of Edge Functions manifest failed
ADDTIONAL PROPERTY must NOT have additional properties

  30 |   ],
  31 |   \\"bundler_version\\": \\"1.6.0\\",
> 32 |   \\"foo\\": \\"bar\\"
     |   ^^^^^ 😲  foo is not expected to be here!
  33 | }"
`;
